tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,498 @@
+  }
+}
+
+template <typename P1, typename P2, typename Index>
+__global__ void PairedSelect(const int num_elements, const P1* p1, const P2* p2,
+                             const Index* indices, P1* selected_p1,
+                             P2* selected_p2) {
+  for (auto idx : CudaGridRangeX(num_elements)) {
+    selected_p1[idx] = p1[indices[idx]];
+    selected_p2[idx] = p2[indices[idx]];
+  }
+}
+
+template <typename T>
+__global__ void Iota(const int num_elements, const T offset, T* to_fill) {
+  for (int idx : CudaGridRangeX(num_elements)) {
+    to_fill[idx] = static_cast<T>(idx) + offset;
+  }
+}
+
+tensorflow::Status NmsGpu(const float* d_desc_sorted_boxes_float_ptr,
+                          const int num_boxes, const float thresh,
+                          int* d_keep_sorted_list, int* h_nkeep,
+                          int* dev_delete_mask, int* host_delete_mask,
+                          OpKernelContext* context, bool flip_boxes) {
+  // d_desc_sorted_boxes_float_ptr is a pointer
+  //    to device memory float array containing the box corners for N boxes.
+  // N is number of boxes.
+  // threshold is the intersection-over-union threshold for elimination
+  // d_keep_sorted_list is a device pointer to int array containing sorted
+  //    indices of the boxes to keep h_nkeep is the pointer to number of
+  //    elements to keep for the host.
+  // h_nkeep is a host pointer for returning number of items to keep.
+  // flip_boxes flag reorders the boxes use lower left and upper right corners
+  // if they are given in mixed format. Making sure we respect the __align(16)__
+  // we promised to the compiler.
+  auto iptr = reinterpret_cast<std::uintptr_t>(d_desc_sorted_boxes_float_ptr);
+  CHECK_EQ((iptr & 15), 0);
+
+  // The next kernel expects squares
+
+  const int mask_ld =
+      (num_boxes + NMS_BOXES_PER_THREAD - 1) / NMS_BOXES_PER_THREAD;
+  const Box* d_desc_sorted_boxes =
+      reinterpret_cast<const Box*>(d_desc_sorted_boxes_float_ptr);
+  auto stream_exec = context->op_device_context()->stream();
+  auto device = context->eigen_gpu_device();
+  dim3 grid, block;
+  int block_size = (num_boxes + NMS_BLOCK_DIM - 1) / NMS_BLOCK_DIM;
+  block_size = std::max(std::min(block_size, NMS_BLOCK_DIM_MAX), 1);
+  grid.x = block_size;
+  grid.y = block_size;
+  block.x = NMS_BLOCK_DIM;
+  block.y = NMS_BLOCK_DIM;
+  if (flip_boxes) {
+    TF_CHECK_OK(GpuLaunchKernel(NMSKernel<true>, grid, block, 0,
+                                device.stream(), d_desc_sorted_boxes, num_boxes,
+                                thresh, mask_ld, dev_delete_mask));
+  } else {
+    TF_CHECK_OK(GpuLaunchKernel(NMSKernel<false>, grid, block, 0,
+                                device.stream(), d_desc_sorted_boxes, num_boxes,
+                                thresh, mask_ld, dev_delete_mask));
+  }
+  CHECK_EQ(cudaGetLastError(), cudaSuccess);
+  // Overlapping CPU computes and D2H memcpy
+  // both take about the same time
```

Comment:
Is this true only for particular architecture? Shall we make NMS_CHUNK_SIZE architecture specific? Please comment.

Reply:
It is more related to PCI-E mem-copy and CPU processing time. So it is somewhat flat nowadays. I wouldn't attack NMS_CHUNK_SIZE first if I am to optimize this code. This is first reference implementation and there are already better kernels. However, considering that a GPU NMS implementation is missing, this would bring significant speed up already. I would prefer to have a working implementation at 90% efficiency now instead of waiting for a 95% efficient version next year. We can improve the kernels at a later time.